src/BayesFilters/src/sigma_point.cpp

```diff
@@ -32,21 +53,263 @@ void bfl::sigma_point::unscented_weights(const unsigned int n, const double alph
 }
 
 
-MatrixXd bfl::sigma_point::sigma_point(const Gaussian& state, const double c)
+MatrixXd bfl::sigma_point::sigma_point(const GaussianMixture& state, const double c)
 {
-    JacobiSVD<MatrixXd> svd = state.covariance().jacobiSvd(ComputeThinU);
+    MatrixXd sigma_points(state.dim, ((state.dim * 2) + 1) * state.components);
 
-    MatrixXd A = svd.matrixU() * svd.singularValues().cwiseSqrt().asDiagonal();
+    for (std::size_t i = 0; i < state.components; i++)
+    {
+        JacobiSVD<MatrixXd> svd = state.covariance(i).jacobiSvd(ComputeThinU);
 
-    MatrixXd sigma_points(state.dim, (state.dim * 2) + 1);
+        MatrixXd A = svd.matrixU() * svd.singularValues().cwiseSqrt().asDiagonal();
 
-    sigma_points << VectorXd::Zero(state.dim), std::sqrt(c) * A, -std::sqrt(c) * A;
+        Ref<MatrixXd> sp = sigma_points.middleCols(((state.dim * 2) + 1) * i, ((state.dim * 2) + 1));
 
-    if (state.dim_linear > 0)
-        sigma_points.topRows(state.dim_linear).colwise() += state.mean().topRows(state.dim_linear);
+        sp << VectorXd::Zero(state.dim), std::sqrt(c) * A, -std::sqrt(c) * A;
 
-    if (state.dim_circular > 0)
-        sigma_points.bottomRows(state.dim_circular) = directional_add(sigma_points.bottomRows(state.dim_circular), state.mean().bottomRows(state.dim_circular));
+        if (state.dim_linear > 0)
+            sp.topRows(state.dim_linear).colwise() += state.mean(i).topRows(state.dim_linear);
+
+        if (state.dim_circular > 0)
+            sp.bottomRows(state.dim_circular) = directional_add(sigma_points.bottomRows(state.dim_circular), state.mean(i).bottomRows(state.dim_circular));
+    }
 
     return sigma_points;
 }
+
+
+std::tuple<bool, GaussianMixture, MatrixXd> bfl::sigma_point::unscented_transform
+(
+    const GaussianMixture& input,
+    const UTWeight& weight,
+    FunctionEvaluation function
+)
+{
+    /* Sample sigma points. */
+    MatrixXd input_sigma_points = sigma_point::sigma_point(input, weight.c);
+
+    /* Propagate sigma points */
+    Data fun_data;
+    bool valid_fun_data;
+    std::tie(valid_fun_data, fun_data) = function(input_sigma_points.cast<float>());
+
+    /* Stop here if function evaluation failed. */
+    if (!valid_fun_data)
+        return std::make_tuple(false, GaussianMixture(), MatrixXd(0, 0));
+
+    /* For now casting Data to MatrixXd. */
+    MatrixXd prop_sigma_points = (bfl::any::any_cast<MatrixXf&&>(std::move(fun_data))).cast<double>();
+
+    /* Initialize transformed gaussian. */
+    GaussianMixture output(input.components, prop_sigma_points.rows());
+
+    /* Initialize cross covariance matrix. */
+    MatrixXd cross_covariance(input.dim, output.dim * output.components);
+
+    /* Process all the components of the mixture. */
+    std::size_t base = ((input.dim * 2) + 1);
+    for (std::size_t i = 0; i < input.components; i++)
+    {
+        Ref<MatrixXd> input_sp = input_sigma_points.middleCols(base * i, base);
```

Comment:
Is there any convenience here in using `Ref<MatrixXd>` instead of `MatrixXd`?

Reply:
The idea is that of using a reference to avoid copying.

 Did you mean if there is convenience in using a `Ref<MatrixXd>` instead of a `MatrixXd&` ?

Reply:
I think there is no copy involved here if you just use `MatrixXd` because there is no actual computation of anything and Eigen optimizes this kind of things. I would go with `MatrixXd` to ease code reading.

Reply:
@claudiofantacci Are you sure that Eigen+compiler are able to optimize those things? Based on my experience, I think it is quite hard for the compiler to optimize away a non-POD local variable  such as that one.

Reply:
Ok, I mark this as resolved after brief discussion with @xEnVrE. Basically you can use Blocks and expressions as lvalue and make operations on them without triggering copies or temporaries provided that you allow the compiler to optimize the code. If used as rvalue, the only way to avoid a copy is by using `Ref`.

Reply:
Exactly, that is what I meant. 

~~~
Ref<MatrixXd> input_sp = input_sigma_points.middleCols(base * i, base); 
input_sp.colwise() -= input.mean(i);
cross_cov.noalias() = input_sp * weight.cov.asDiagonal() * prop_sp.transpose();
~~~

is equivalent to (I guess): 
~~~
cross_cov.noalias() = ( input_sigma_points.middleCols(base * i, base).colwise() -  input.mean(i) ) * weight.cov.asDiagonal() * prop_sp.transpose();
~~~
thanks to Eigen expressions bla bla, but this cannot be substituted with: 
~~~
Eigen::MatrixXd input_sp = input_sigma_points.middleCols(base * i, base); 
input_sp.colwise() -= input.mean(i);
cross_cov.noalias() = input_sp * weight.cov.asDiagonal() * prop_sp.transpose();
~~~ 
as the "expression" returned by `middleCols` will be evaluated in the constructor of `MatrixXd`.

Reply:
Yes you were right! 👍